Gitops Example: Supply Chain to Delivery

waciumawanjohi opened the pull request:
---
name: Gitops Example: Supply Chain to Delivery
about: Updating the example

---
<!--
Thanks for submitting a pull request to Cartographer!

Also check the [PR guidelines] if you haven't already!
-->

[PR requirements]: https://github.com/vmware-tanzu/cartographer/blob/main/CONTRIBUTING.md#commit-message-and-pr-guidelines

## Changes proposed by this PR

<!--
Add the story that is resolved or updated
`closes`/`fixes` or `updates` are valid here 
-->
closes #405

<!--
Summarize your changes. Please include reasoning and key decisions to help the reviewer
understand the changes.
-->


## Release Note

<!--
Your PR title will be directly included in the release notes when it ships in
the next release. It should briefly describe the PR in [imperative
mood]. Please refrain from adding prefixes like 'feature:', and don't include a
period at the end.

Within this section you may supply a list of extra information to include in
the release notes in addition to the pull request title.

Example title: Add CreatorName field in Workload spec

Example notes:

* Operators can label stamped objects with the Workload's creator
* Creators can be contacted when interdepartmental issues arise 

If there are no additional notes necessary you may remove this entire section.
-->
[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative
- Adds example of a supply chain that ends by writing app config to git
- Adds example of a delivery that reads that git repo and deploys the app in cluster
- Adds verification of these examples to the end-to-end test

## PR Checklist

Note: Please do not remove items. Mark items as done `[x]` or use ~strikethrough~ if you believe they are not relevant

- [x] Linked to a relevant issue. Eg: `Fixes #123` or `Updates #123`
- [ ] Removed non-atomic or `wip` commits
- [x] Filled in the [Release Note](#Release-Note) section above 
- [x] Modified the docs to match changes <!-- TBD: reference doc editing guidance -->

waciumawanjohi commented:
This PR only addresses the multi-cluster, gitops, supply-chain => delivery flow. There are comments in #405 about including an example of image scanning. If that's desired, we should write a separate story for that.

martyspiewak commented on examples/README.md:
```diff
@@ -1,6 +1,23 @@
 # Examples
```
We should probably add the delivery and gitops examples to this README as well

martyspiewak commented on examples/basic-delivery/README.md:
````diff
@@ -0,0 +1,338 @@
+The App can be inspected for objects created:
+
+```bash
+kgy app gitops | yq -r .status.inspect.stdout
````
I assume this is an alias that accidentally made it in here?
```suggestion
kubectl get app gitops -o yaml | yq -r .status.inspect.stdout
```

martyspiewak commented on hack/setup.sh:
```diff
@@ -248,49 +354,296 @@ install_tekton() {
+        export test_name="$1"
+        kapp delete --yes -a "example-$test_name"
+        kapp delete --yes -a "setup-example-$test_name"
+        until [[ -z $(kubectl get pods -l "serving.knative.dev/configuration=$test_name" -o name) ]]; do sleep 1; done
```
This line has been hanging for me because that pod is not getting deleted and it is stopping the tests from progressing. Is this something wrong on my end (I don't think it is because I don't think the deletes above should remove this pod but I could definitely be wrong)?

waciumawanjohi replied:
This succeeds for me on both my personal machine and on a linux VM. But it fails in the github runner. I've commented it out for now.

martyspiewak commented on hack/setup.sh:
```diff
@@ -61,19 +78,36 @@ main() {
+                        teardown_runnable_example
+                        teardown_example_sc "basic-sc"
+                        teardown_example_sc "testing-sc"
+                        teardown_gitops_example
```
`teardown_gitops_example` is already called in `test_gitops` so I think it'd end up being called twice if we leave this here. Maybe it should only be here and we shouldn't teardown in `test_gitops`

waciumawanjohi replied:
This is true for the other tests as well. If they succeed, they clean up after themselves. If everything has succeeded, teardown-example needn't be called.

So why have any of the tests clean up? Shortcut to avoid test pollution.

martyspiewak commented on hack/setup.sh:
```diff
@@ -46,6 +62,7 @@ main() {
                 case $command in
                 cluster)
                         start_registry
+                        start_repository
```
I'm not sure I'd want a repository and gitea to be started everything I use this script to start a local cluster with cartographer.

I'd be in favor of making the gitea and gitops setup a separate step (and making the gitops testing separate as well) as they are a bit heavier to test and require more setup that we may not want to go through every time we are running tests locally. 

If not, at the very least I think we should defer this until `example-dependencies` so that the repository isn't setup unless you are setting up the examples. But as I said, I think I'd prefer to have the gitops setup and testing as a separate command. 

What do you think?

waciumawanjohi replied:
I moved `start_repository` to `example_dependencies`. I've kept all the tests together. My expectation is that the examples are run by the end to end tests, and that devs aren't running that e2e on their local machines.

waciumawanjohi committed:
Example of Delivery

Create example of supply chain that concludes by writing to git.
Create an example of a Delivery that picks up the configuration
written to git and applies that config to the cluster (i.e.
deploys the app).

martyspiewak reviewed (approved):
lgtm!
